com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/UploadAppStep.java

```diff
@@ -106,28 +138,30 @@ private UploadToken asyncUploadFiles(ExecutionWrapper execution, CloudController
         return uploadToken;
     }
 
-    protected Path extractFromMtar(InputStream appArchiveStream, String fileName, long maxSize) {
-        ApplicationArchiveExtractor appExtractor = new ApplicationArchiveExtractor(appArchiveStream, fileName, maxSize, getStepLogger());
-        return appExtractor.extract();
+    protected ApplicationArchiveExtractor getApplicationArchiveExtractor(InputStream appArchiveStream, String fileName, long maxSize,
+        Set<String> knownFileNames) {
+        return new ApplicationArchiveExtractor(appArchiveStream, fileName, maxSize, knownFileNames, getStepLogger());
+    }
+
+    protected Path extractFromMtar(ApplicationArchiveExtractor appExtractor) {
+        return appExtractor.extractApplicationInNewArchive();
     }
 
     private void upload(ExecutionWrapper execution, CloudControllerClient client, CloudApplication app, Path filePath,
-        UploadToken uploadToken) throws IOException {
-        detectApplicationFileDigestChanges(execution, app, filePath.toFile(), client);
-        UploadToken currentUploadTokent = client.asyncUploadApplication(app.getName(), filePath.toFile(),
-            getMonitorUploadStatusCallback(app, filePath.toFile(), execution.getContext()));
-        uploadToken.setPackageGuid(currentUploadTokent.getPackageGuid());
-        uploadToken.setToken(currentUploadTokent.getToken());
+        UploadToken uploadToken, CloudResources knownRemoteResources) throws IOException {
+        UploadToken currentUploadToken = client.asyncUploadApplication(app.getName(), filePath.toFile(),
+            getMonitorUploadStatusCallback(app, filePath.toFile(), execution.getContext()), knownRemoteResources);
+        uploadToken.setPackageGuid(currentUploadToken.getPackageGuid());
+        uploadToken.setToken(currentUploadToken.getToken());
     }
 
-    private void detectApplicationFileDigestChanges(ExecutionWrapper execution, CloudApplication app, File applicationFile,
+    private void detectApplicationFileDigestChanges(ExecutionWrapper execution, CloudApplication app, String newApplicationDigest,
         CloudControllerClient client) {
         CloudApplication existingApp = client.getApplication(app.getName());
         ApplicationFileDigestDetector applicationFileDigestDetector = new ApplicationFileDigestDetector(existingApp);
-        String appNewFileDigest = applicationFileDigestDetector.detectNewAppFileDigest(applicationFile);
         String currentFileDigest = applicationFileDigestDetector.detectCurrentAppFileDigest();
-        attemptToUpdateApplicationDigest(client, app, appNewFileDigest, currentFileDigest);
-        setAppContentChanged(execution, hasAppFileDigestChanged(appNewFileDigest, currentFileDigest));
+        attemptToUpdateApplicationDigest(client, app, newApplicationDigest, currentFileDigest);
+        setAppContentChanged(execution, hasAppFileDigestChanged(newApplicationDigest, currentFileDigest));

```

Comment:
Be sure that when the application binaries are not changed, the application will not be restarted as this digest is handling this case :)